High output management pt1 (#2)
vlad17/books: high-output-management/chapters1-2.tex

\documentclass{article}

\title{Part 1: The Basics of Production\\Chapters 1 to 2}
\author{Vladimir Feinberg}

\input{../defs}

\begin{document}

\maketitle

Notes on Chapters 1 to 2 of Andy Grove's \textit{High Output Management} \cite{highoutput}.

Note that the sections are structured according to topics, not the chapters themselves

\section{Creating Production Requirements}

Work backwards from what the market needs. What do customers desire?

\begin{enumerate}
\item Timely delivery
\item High quality
\item Low cost
\end{enumerate}

> Then the question is, how do you trade off each of the above? At Intel, the tradeoff seemed to be

Reviewer: ```suggestion
Then the question is, how do you trade off each of the above? At Intel, the trade-off seemed to be
```

\begin{align*} 
  \min \,\,\,& \texttt{cost}\\
  \text{s.t.}\,\,\, & \texttt{quality}\ge \text{some acceptable threshold} \\
 & \texttt{delivery time}\le \text{sufficiently timely}
\end{align*}

> But how should you set up the optimization problem in the context of your business? Andy does not answer this question for you. I suspect that the decision to, e.g., minimize cost subject to thresholds on quality and delivery time instead of minimize delivery time subject to cost and quality constraints (or some completely different optimization resulting from different latent objectives) is one of a general business model. For instance, for Amazon prime, delivery time is much more important.

Reviewer: can you change the "I suspect..." sentence it's a little wordy and hard to interpret.

Author: lmk if addr now

Consider the following approaches.
\begin{enumerate}

> \item Profile your customer to see what they care about more. For instances, in some industries, like ecommerce, cost is actually not as important as fast delivery turnaround (maybe, just guessing).

Reviewer: ```suggestion
\item Profile your customer to see what they care about more. For instance, in some industries, like e-commerce, cost is actually not as important as fast delivery turnaround (maybe, just guessing).
```

Reviewer: That statement is true up to a value. You can offload costs on to customer, but also give them cheaper option. For example, Amazon gives you a gift card to use their products if you choose the slowest delivery option.

\item It just comes from the industry, market, and your business model or unique approach. This is what makes your business special: it's your business model
\end{enumerate}

\section{Production planning}

The key insight is to \textit{consider planning tasks with critical-path planning in the workflow DAG}. Work backwards from delivery time and requirements to construct the DAG of tasks to perform, and track the critical path progress over time.

Don’t forget to incorporate non-productive tasks into your planning.
\begin{enumerate}
\item Assembly
\item Testing
\item Inspection
\item Queue wait time in components that have limited capacity
\end{enumerate}

> The above planning informs inventory purchases. For example, in a development environment, if your Travis CI only handles 1 job at a time then you can’t merge in more than one PR every 30 minutes or however long your build time is. In turn, you should never bottleneck yourself on CI by buying more concurrent builds as you get more developers.

Reviewer: ```suggestion
The above planning informs inventory purchases. For example, in a development environment, if your Travis Continuous Integration only handles 1 job at a time then you can’t merge in more than one PR every 30 minutes or however long your build time is. In turn, you should never bottleneck yourself on CI by buying more concurrent builds as you get more developers.
```

\item In software, consider using both the number of PRs (``diffs'' or ``branches'') merged as well as per-PR SLOC as a measure of developer productivity (neither is perfectly accurate, but both should be considered).
\end{itemize}

Provide variance estimates on measurements. Not every number should be taken under equal consideration. With variance estimates, you can add slack to scale with desired confidence in allocations. Critical processes may need more slack.

In trend prediction,\footnote{This section is pretty weird, kind of like Andy just teaching you forecasting.} make sure that you use:
\begin{itemize}
\item Historical data
\item Capture seasonality
\item Autoregressive trends
\item Nonlinearity (make sure you know what kind you need to have in your models)
\end{itemize}

Properly taking advantage of trends can lead to a lot of value creation (see the \nurl{https://www.cnbc.com/2018/05/03/how-ray-dalio-helped-launch-mcdonalds-chicken-mcnugget.html}{Bridgewater example}).

\section{De-risking Parallel Flows}

> If you have two dependent flows that are both deep and need to converge to create value, then you're undertaking a lot of risk. For instance, consider some manufacturing process that transforms raw material into goods and an analogous sales process that are both required to start early (for long sales cycles) that result into a successful deal (see p.26 of the book for a good picture).

Reviewer: Can you find and insert the picture here.

> Fail early. Capture failures at the lowest value stage (earlier) that is possible to detect them (needs to be traded off with detection costs). For instance, in candidate recruitment: assess candidate interest early in the pipeline, and if they’re just shopping around then they’re not worth the interview time investment.

Reviewer: What do u mean by early in the pipeline. Some examples are like phone screen or on-campus interviews to reduce cost.

Author: Added `(i.e., during initial phone screens)`

> Introduce synchronization points as much as possible between parallel flows. Two parallel but dependent features that two teams are working should be incrementalized and interacting at as early stages as possible (rather than build two things separately then connect). An example in code development would be to rebase often.

Reviewer: ```suggestion
Introduce synchronization points as much as possible between parallel flows. Two parallel but dependent features that two teams are working should be incrementalized and should be interacting at as early stages as possible (rather than build two things separately then connect). An example in code development would be to rebase often.
```

> Introduce synchronization points as much as possible between parallel flows. Two parallel but dependent features that two teams are working should be incrementalized and interacting at as early stages as possible (rather than build two things separately then connect). An example in code development would be to rebase often.

Reviewer: ```suggestion
Introduce synchronization points as much as possible between parallel flows. Two parallel but dependent features that two teams are working should be incrementalized and interacting at as early stages as possible (rather than building two things separately, then connect). An example in code development would be to rebase often.
```

Program slack in volumes of output so that failures downstream can be compensated for. Deadlines for a feature from, say, a customer request, should be made with slack.

Introduce additional quality assurance to take into account the additional risk that the parallel flows incur whenever cheap, continuous tests are amenable (need to trade off w/ quality assurance risk).

\section{Quality Assurance}

\subsection{Gated Tests}

Gated tests, which “hold-the-line” until the completely check out a batch of a product, or take a while to decisively complete, but don’t allow further changes to be made. In software:

\begin{itemize}
\item Diff review
\item Pen test
\item CI is a very cheap gated test
\end{itemize}

\subsection{Sampling-based Tests}

> Sampling-based tests probibalistically evaluate for potential failures by taking look at only parts of the actual total amount of product you’re producing. If they fail then they indicate there’s a broader problem

Reviewer: ```suggestion
Sampling-based tests probabilistically evaluate for potential failures by taking look at only parts of the actual total amount of product you’re producing. If they fail then they indicate there’s a broader problem
```

> Sampling-based tests probibalistically evaluate for potential failures by taking look at only parts of the actual total amount of product you’re producing. If they fail then they indicate there’s a broader problem

Reviewer: ```suggestion
Sampling-based tests probibalistically evaluate for potential failures by taking a look at only parts of the actual total amount of product you’re producing. If they fail then they indicate there’s a broader problem
```
